superset/views/base.py

```diff
@@ -327,6 +329,17 @@ def common_bootstrap_payload() -> Dict[str, Any]:
     }
 
 
+def get_moment_locale(locale):
+    moment_locale = locale
+    if "_" in moment_locale:
+        moment_locale = moment_locale.split("_")[0]
+
+    flag = conf.get("LANGUAGES").get(locale).get("flag")
```

Comment:
`flag` is for the country flag icon, which doesn't seem right to make it as a dependency for locale inference.

You should probably use [`babel.core.negotiate_locale`](http://babel.pocoo.org/en/latest/api/core.html#babel.core.negotiate_locale) to select from `humanize`'s [supported locales](https://github.com/jmoiron/humanize/tree/master/src/humanize/locale):

```python
locale = negotiate_locale([str(get_locale())], ['de_DE', 'es_ES', 'fa_IR', 'fi_FL'...])
```

It seems `humanize` doesn't export the list of supported locales so you may need to manually define this list as a constant.

Reply:
@ktmud  
`humanize` and `moment.js` and `superset` use different locale keys. I didn't add variables by myself, so I only get variables in the current `LANGUAGES` config.
Do we need to expand the current `LANGUAGES` config? Just like this:
```python
LANGUAGES = {
    "en": {"flag": "us", "name": "English", "moment_locale": "en", "humanize_locale": "en", "d3_time_locale": "en-US"},
    "it": {"flag": "it", "name": "Italian", "moment_locale": "it", "humanize_locale": "en", "d3_time_locale": "it-IT"},
    "fr": {"flag": "fr", "name": "French", "moment_locale": "fr", "humanize_locale": "fr_FR", "d3_time_locale": "fr-FR"},
    "zh": {"flag": "cn", "name": "Chinese", "moment_locale": "zh-cn", "humanize_locale": "zh_CN", "d3_time_locale": "zh-CN"},
    "ja": {"flag": "jp", "name": "Japanese", "moment_locale": "ja", "humanize_locale": "jp", "d3_time_locale": "ja-JP"},
    "de": {"flag": "de", "name": "German", "moment_locale": "de", "humanize_locale": "de_DE", "d3_time_locale": "de-DE"},
    "pt": {"flag": "pt", "name": "Portuguese", "moment_locale": "pt", "humanize_locale": "en", "d3_time_locale": "en-US"},
    "pt_BR": {"flag": "br", "name": "Brazilian Portuguese", "moment_locale": "pt-br", "humanize_locale": "pt_BR", "d3_time_locale": "pt-BR"},
    "ru": {"flag": "ru", "name": "Russian", "moment_locale": "ru", "humanize_locale": "ru_RU", "d3_time_locale": "ru-RU"},
    "ko": {"flag": "kr", "name": "Korean", "moment_locale": "ko", "humanize_locale": "ko_KR", "d3_time_locale": "ko-KR"},
}
```

Reply:
Oh~ I forgot to mention that there is actually a separate locale definition value for time formatting in `d3-format`, I've appended each value.

Reply:
I'd prefer not to add them in the configs as the languages are enumerable and there seems to be an easy way to transform between each. You may create an util function on both backend and frontend to consolidate the format for each library.